File: src/client/java/io/github/romeoahmed/cursedoath/client/render/domain/ShrineSlashes.java
package io.github.romeoahmed.cursedoath.client.render.domain;

import io.github.romeoahmed.cursedoath.client.render.EffectMesh;
import it.unimi.dsi.fastutil.HashCommon;
import net.minecraft.core.BlockPos;
import net.minecraft.world.phys.Vec3;

/// Decorative cuts stay anchored to domain-local cells as the camera crosses cell boundaries.
final class ShrineSlashes {
    private static final long BYTE_MASK = 255L;
    private static final double JITTER = 0.6;
    private static final double CELL_INSET = 0.2;
    private static final double CELL = 8.0;
    private static final int REACH = 4;
    private static final float PERIOD = 10f;
    private static final float LIFETIME = 4f;
    private static final float REVEAL = 0.65f;
    private static final double NEAR = 3.0;
    private static final double CLEAR = 7.0;
    private static final double FAR = 32.0;
    private static final double FADE = 8.0;
    private static final double WIDTH = 0.035;
    private static final int COLOR = 0xE6D7CF;
    private static final int ACCENT_EVERY = 7;
    private static final double ACCENT_LENGTH = 9.0;
    private static final double LENGTH = 5.0;
    private static final float SUBDUED_ALPHA = 0.35f;
    private static final double PHASE_STEP = 0.61803398875;
    private static final double TURN = 6.28318530718;
    private final EffectMesh mesh;
    private final double radius;
    private final float age;
    private final Vec3 eye;
    private final boolean subdued;

    ShrineSlashes(EffectMesh mesh, double radius, float age, Vec3 eye, boolean subdued) {
        this.mesh = mesh;
        this.radius = radius;
        this.age = age;
        this.eye = eye;
        this.subdued = subdued;
    }

    void draw() {
        int cx = (int) Math.floor(eye.x / CELL),
                cy = (int) Math.floor(eye.y / CELL),
                cz = (int) Math.floor(eye.z / CELL);
        for (int x = cx - REACH; x <= cx + REACH; x++)
            for (int y = cy - REACH; y <= cy + REACH; y++) for (int z = cz - REACH; z <= cz + REACH; z++) cut(x, y, z);
    }

    private void cut(int x, int y, int z) {
        long seed = HashCommon.mix(BlockPos.asLong(x, y, z));
        if (subdued && (seed & 1) == 0) return;
        float offset = (float) ((seed >>> 16) % 1000) / 1000 * PERIOD, clock = age + offset;
        if (clock % PERIOD >= LIFETIME) return;
        int beat = (int) Math.floor(clock / PERIOD);
        double phase = (double) (seed & 0xFFFF) / 65536 * TURN + beat * PHASE_STEP;
        double length = (seed + beat) % ACCENT_EVERY == 0 ? ACCENT_LENGTH : LENGTH;
        var center = new Vec3(
                (x + offset(seed >>> 24)) * CELL, (y + offset(seed >>> 32)) * CELL, (z + offset(seed >>> 40)) * CELL);
        if (eye.distanceToSqr(center) >= FAR * FAR || center.length() + length > radius) return;
        var direction = new Vec3(Math.cos(phase), Math.sin(phase * 2), Math.sin(phase)).normalize();
        var delta = direction.scale(length);
        var a = center.subtract(delta);
        var b = center.add(delta);
        float alpha = visibility(eye, a, b) * (subdued ? SUBDUED_ALPHA : 1);
        if (alpha <= 0) return;
        float progress = clock % PERIOD / LIFETIME;
        var tip = a.lerp(b, Math.min(clock % PERIOD / REVEAL, 1f));
        var width = direction.cross(eye.subtract(center)).normalize().scale(WIDTH);
        mesh.slash(a, tip, width, COLOR, alpha * (1 - progress));
    }

    private static float visibility(Vec3 eye, Vec3 a, Vec3 b) {
        var motion = b.subtract(a);
        double fraction = Math.clamp(eye.subtract(a).dot(motion) / motion.lengthSqr(), 0, 1),
                nearest = eye.distanceTo(a.lerp(b, fraction)),
                distance = eye.distanceTo(a.lerp(b, 0.5));
        return (float)
                (Math.clamp((nearest - NEAR) / (CLEAR - NEAR), 0, 1) * Math.clamp((FAR - distance) / FADE, 0, 1));
    }

    private static double offset(long seed) {
        return (double) (seed & BYTE_MASK) / BYTE_MASK * JITTER + CELL_INSET;
    }
}

File: src/client/kotlin/io/github/romeoahmed/cursedoath/client/render/domain/ShrineSlashes.kt
package io.github.romeoahmed.cursedoath.client.render.domain

import io.github.romeoahmed.cursedoath.client.render.EffectMesh
import it.unimi.dsi.fastutil.HashCommon
import net.minecraft.core.BlockPos
import net.minecraft.world.phys.Vec3
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.sin

/** Domain-local cells keep cuts anchored when the observer crosses a sampling boundary. */
internal class ShrineSlashes(
    private val mesh: EffectMesh,
    private val radius: Double,
    private val age: Float,
    private val eye: Vec3,
    private val subdued: Boolean,
) {
    fun draw() {
        val cx = floor(eye.x / CELL).toInt()
        val cy = floor(eye.y / CELL).toInt()
        val cz = floor(eye.z / CELL).toInt()
        for (x in cx - REACH..cx + REACH) {
            for (y in cy - REACH..cy + REACH) {
                for (z in cz - REACH..cz + REACH) {
                    cut(x, y, z)
                }
            }
        }
    }

    private fun cut(
        x: Int,
        y: Int,
        z: Int,
    ) {
        val seed = HashCommon.mix(BlockPos.asLong(x, y, z))
        if (subdued && seed and 1L == 0L) return
        val offset = ((seed ushr 16) % 1000).toFloat() / 1000 * PERIOD
        val clock = age + offset
        if (clock % PERIOD >= LIFETIME) return
        val beat = floor(clock / PERIOD).toInt()
        val phase = (seed and 0xFFFF).toDouble() / 65536 * TURN + beat * PHASE_STEP
        val length = if ((seed + beat) % ACCENT_EVERY == 0L) ACCENT_LENGTH else LENGTH
        val center =
            Vec3(
                (x + offset(seed ushr 24)) * CELL,
                (y + offset(seed ushr 32)) * CELL,
                (z + offset(seed ushr 40)) * CELL,
            )
        if (eye.distanceToSqr(center) >= FAR * FAR) return
        // Keep the complete cut inside the sure-hit volume, including its pointed tips.
        if (center.length() + length > radius) return
        val direction = Vec3(cos(phase), sin(phase * 2), sin(phase)).normalize()
        val delta = direction.scale(length)
        val a = center.subtract(delta)
        val b = center.add(delta)
        val alpha = visibility(eye, a, b) * if (subdued) SUBDUED_ALPHA else 1f
        if (alpha <= 0f) return
        val progress = clock % PERIOD / LIFETIME
        val tip = a.lerp(b, (clock % PERIOD / REVEAL).coerceAtMost(1f).toDouble())
        val width = direction.cross(eye.subtract(center)).normalize().scale(WIDTH)
        mesh.slash(a, tip, width, COLOR, alpha * (1 - progress))
    }

    private fun visibility(
        eye: Vec3,
        a: Vec3,
        b: Vec3,
    ): Float {
        val motion = b.subtract(a)
        val fraction = (eye.subtract(a).dot(motion) / motion.lengthSqr()).coerceIn(0.0, 1.0)
        val nearest = eye.distanceTo(a.lerp(b, fraction))
        val distance = eye.distanceTo(a.lerp(b, 0.5))
        return (
            ((nearest - NEAR) / (CLEAR - NEAR)).coerceIn(0.0, 1.0) *
                ((FAR - distance) / FADE).coerceIn(0.0, 1.0)
        ).toFloat()
    }

    private fun offset(seed: Long): Double = (seed and BYTE_MASK).toDouble() / BYTE_MASK * JITTER + CELL_INSET

    private companion object {
        const val BYTE_MASK = 255L
        const val JITTER = 0.6
        const val CELL_INSET = 0.2
        private const val CELL = 8.0
        private const val REACH = 4
        private const val PERIOD = 10f
        private const val LIFETIME = 4f
        private const val REVEAL = 0.65f
        private const val NEAR = 3.0
        private const val CLEAR = 7.0
        private const val FAR = 32.0
        private const val FADE = 8.0
        private const val WIDTH = 0.035
        private const val COLOR = 0xE6D7CF
        private const val ACCENT_EVERY = 7
        private const val ACCENT_LENGTH = 9.0
        private const val LENGTH = 5.0
        private const val SUBDUED_ALPHA = 0.35f
        private const val PHASE_STEP = 0.61803398875
        private const val TURN = 6.28318530718
    }
}
